### DISCUSSION TITLE
Miryoku x Keyball

### DISCUSSION THREAD

**Discussion by sohrab-** [manna-harbour/miryoku#309] [Q&A] [answered]
I am trying to apply Miryoku to [Keyball39](https://github.com/Yowkees/keyball/tree/main/qmk_firmware/keyboards/keyball/keyball39). It supports QMK and provides a couple of layouts out of the box. 

Is there a way to readily apply Miryoku to this keyboard? (Sorry if this is a dumb question but I am still trying to get my head around QMK.)

As far as I can tell, it doesn't support community layouts or use the new `info.json` stuff.

If this requires some legwork, I am happy to do the work to get the layout integrated but I probably need some pointers on the approach.

**Comment by manna-harbour** [comment, marked as answer]
First you'll need to merge the keyboard definition with Miryoku QMK. Ordinarily that could be done with the workflow automatic merge function, but at the moment it will need to be copied manually. See https://github.com/manna-harbour/miryoku/discussions/287. 

Other than that you need to add a mapping to Miryoku QMK. See https://github.com/manna-harbour/miryoku/discussions/264#discussioncomment-7689140. Refer to https://github.com/Yowkees/keyball/blob/98595e59944192ab47df3936b4b892a5f15bc846/qmk_firmware/keyboards/keyball/keyball39/keymaps/default/keymap.c#L27-L30.

Since there are only 2 thumb keys on the ball side you'll also need to enable thumb combos.

Mappings can also be added by request.

**Reply by sohrab-** [reply]
Thanks for the pointers. 

Here is my `keyball39/keymaps/manna-harbour_miryoku/config.h` if it helps anyone else.

```c
#pragma once

#define XXX KC_NO

#define LAYOUT_miryoku( \
K00,   K01,   K02,   K03,   K04,          K05,   K06,   K07,   K08,   K09, \
K10,   K11,   K12,   K13,   K14,          K15,   K16,   K17,   K18,   K19, \
K20,   K21,   K22,   K23,   K24,          K25,   K26,   K27,   K28,   K29, \
N30,   N31,   K32,   K33,   K34,          K35,   K36,   K37,   N38,   N39 \
) \
LAYOUT( \
K00,   K01,   K02,   K03,   K04,          K05,   K06,   K07,   K08,   K09, \
K10,   K11,   K12,   K13,   K14,          K15,   K16,   K17,   K18,   K19, \
K20,   K21,   K22,   K23,   K24,          K25,   K26,   K27,   K28,   K29, \
XXX,   XXX,   XXX,   K32,   K33,   K34,   K35,   K36,   XXX \
)

```

(Requires `#define LAYER_STATE_32BIT`.)

**Reply by nomm-dot** [reply]
Would you be willing to give an explanation of how you got the miryoku layout onto your keyball39? Just got one myself, and got it up in the REMAP editor after some difficulty, but have not been able to find any resources on how to get a different keymap on REMAP and every resource I can find for editing the layout of the keyball series says you have to use REMAP. I am also trying to wrap my head around how QMK works as well as the keyball. Some insight on how you got yours working would be great.

**Reply by sohrab-** [reply]
I followed the steps here: https://github.com/Yowkees/keyball/blob/main/qmk_firmware/keyboards/keyball/readme.md#how-to-build

Then I added my keymap under `qmk_firmware/keyboards/keyball/keyball39/keymaps/manna-harbour_miryoku`, with `config.h` above, along with the following files:

`keymap.c`:

```c
#include QMK_KEYBOARD_H

#include "quantum.h"
```

`rules.mk`:

```make
RGBLIGHT_ENABLE = no

OLED_ENABLE = yes
```

And ran the build per instructions, in my case:

```
make SKIP_GIT=yes clean keyball/keyball39:manna-harbour_miryoku MIRYOKU_ALPHAS=QWERTY MIRYOKU_CLIPBOARD=MAC MIRYOKU_KLUDGE_THUMBCOMBOS=yes
```

Then I flash my keyboard (I use QMK toolbox but can be done from CLI too) using the generated `qmk/keyball_keyball39_manna-harbour_miryoku.hex` file.

**Reply by nomm-dot** [reply]
Thank you, this helped a lot. I copied your config.h into the custom_config.h, and changed the RGBLIGHT_ENABLE to yes in custom_rules.mk. I ran your provided make command with slight modifications and got an error claiming it could not find a keymap, but it seems to be referencing the keymap in the keyball39 dir instead of the miryoku keymap.

```
[noah@Framework:~/qmk_firmware]$ make SKIP_GIT=yes clean keyball/keyball39:manna-harbour_miryoku MIRYOKU_ALPHAS=QWERTY MIRYOKU_NAV=INVERTEDT MIRYOKU_KLUDGE_THUMBCOMBOS=yes 
Deleting .build/ ... done.
☒ Not including data from file: keyboards/keyball/keyball39/info.json
☒       rgblight: Additional properties are not allowed ('driver' was unexpected)☒ keyball/keyball39: No LAYOUTs defined! Need at least one layout defined in info.json.
Making keyball/keyball39 with keymap manna-harbour_miryoku

☒ Not including data from file: keyboards/keyball/keyball39/info.json
☒       rgblight: Additional properties are not allowed ('driver' was unexpected)☒ keyball/keyball39: No LAYOUTs defined! Need at least one layout defined in info.json.

 * Could not find keymap                                                                               [ERRORS]

builddefs/build_keyboard.mk:157: *** Invalid keymap.  Stop.
Make finished with errors
make: *** [Makefile:389: keyball/keyball39:manna-harbour_miryoku] Error 1

```

I've got the my directory structure the same as yours `~/qmk_firmware/keyboards/keyball/keyball39/keymaps/manna-harbour_miryoku`

**Reply by sohrab-** [reply]
Did you symlink per the doc?

**Reply by nomm-dot** [reply]
I did, I believe I did it correctly. I don't use symlinks too often. here's a screenshot of what it looks like in the terminal.
![2024-07-15-13:05:52](https://github.com/user-attachments/assets/17dd4653-c2a4-47e0-b952-cfb99afcf935)

and here is the command I ran from the ~/qmk_firmware directory. I also commented out the OLED and RGB_ENABLE, but that did not seem to change the output at all.
![2024-07-15-13:10:22](https://github.com/user-attachments/assets/836cf60f-1f45-4270-bb06-befd0f7c3c2e)

Additionally, this may or may not be common or relevant, but there's a qmk_firmware directory inside of the keyball directory. This seems strange to me.
![2024-07-15-13:20:48](https://github.com/user-attachments/assets/d9410a29-99cf-4d2a-93ed-2612c743d461)

**Reply by nomm-dot** [reply]
Would you be able to make keyball39 miryoku config available in a repo so that I can compare it to what I have? I started from scratch, but still getting the same errors. I must be missing obvious, or something that will be difficult to get assistance with over a format like this. I've only used qmk and zmk through github actions, so my basic understanding at this point is very minimal.

**Reply by sohrab-** [reply]
Yeah, this looks ok to me. If inside the QMK cloned repo, you can do this, then the symlink is done correctly:

```
λ ls keyboards/keyball/keyball39/keymaps/manna-harbour_miryoku
config.h keymap.c rules.mk
```

I do also note I am 0.22.14 git tag of QMK repo.

The `qmk_firmware` inside `keyball` repo is fine. That's what you are symlinking if you look at your previous screenshot.

I haven't made any other changes but here is my git patch anyway if this helps: [keyball.patch](https://github.com/user-attachments/files/16241517/keyball.patch)

**Reply by nomm-dot** [reply]
@sohrab-  I believe I might have finally gotten close by setting up a new environment in a different distro. I'm getting an error that describes the maximum layer limit is exceeded. I see in your config.h reply that you specified `#define LAYER_STATE_32BIT` is required, but I'm not sure where to declare that I've tried putting it in my config.h and specifying LAYER_STATE_32BIT in when trying to build the hex, but I continue to get the error. Recommendations? I can't seem to find this in the QMK documentation.

**Reply by nomm-dot** [reply]
Ah, I see, I changed it in `keyball/qmk_firmware/keyboards/keyball/keyball39/config.h`, but now my firmware is too large. Did you manage to get around this somehow?
